File: MileageStats.Domain/Models/Vehicle.cs
/*  
Copyright Microsoft Corporation

Licensed under the Apache License, Version 2.0 (the "License"); you may not
use this file except in compliance with the License. You may obtain a copy of
the License at 

http://www.apache.org/licenses/LICENSE-2.0 

THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED 
ARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE, 
MERCHANTABLITY OR NON-INFRINGEMENT. 

See the Apache 2 License for the specific language governing permissions and
limitations under the License. */

using System.Collections.Generic;
using MileageStats.Domain.Contracts;

namespace MileageStats.Domain.Models
{
    /// <summary>
    /// A vehicle used to track mileage statistics.
    /// </summary>
    public class Vehicle :IHasIdentity
    {
        public Vehicle()
        {
            Reminders = new List<Reminder>();
        }

        public int Id { get { return VehicleId; } set { VehicleId = value; } }

        /// <summary>
        /// Gets or sets the entity ID of vehicle.
        /// </summary>
        /// <value>
        /// An integer identifying the entity.
        /// </value>
        public int VehicleId { get; set; }

        /// <summary>
        /// Gets or sets the entity ID of user who owns the vehicle.
        /// </summary>
        /// <value>
        /// An integer identifying the entity.
        /// </value>
        public int UserId { get; set; }

        /// <summary>
        /// Gets or sets the name of the vehicle.
        /// </summary>
        /// <value>
        /// A string.
        /// </value>
        public string Name { get; set; }

        /// <summary>
        /// Gets or sets the sort order relative to other vehicles.
        /// </summary>
        /// <value>
        /// A positive number or zero.
        /// </value>
        public int SortOrder { get; set; }

        /// <summary>
        /// Gets or sets the photo of the vehicle.
        /// </summary>
        /// <value>
        /// The photo.
        /// </value>
        public VehiclePhoto Photo { get; set; }

        /// <summary>
        /// Gets or sets the entity ID of photo for the vehicle.
        /// </summary>
        /// <value>
        /// An integer identifying the entity.
        /// </value>
        public int PhotoId { get; set; }

        /// <summary>
        /// Gets or sets the manufacturing year of the vehicle.
        /// </summary>
        /// <value>
        /// An integer after 1896.
        /// </value>        
        public int? Year { get; set; }

        /// <summary>
        /// Gets or sets the make of the vehicle (e.g. Toyota, Ford).
        /// </summary>
        /// <value>
        /// A string.
        /// </value>
        public string MakeName { get; set; }

        /// <summary>
        /// Gets or sets the model of the vehicle (e.g. Camry, Fiesta)
        /// </summary>
        /// <value>
        /// A string.
        /// </value>
        public string ModelName { get; set; }

        /// <summary>
        /// The reminders for this vehicle.  Depending on the method of retrieval from the repository, this property may not necessarily get filled.
        /// </summary>
        public ICollection<Reminder> Reminders { get; set; }
    }
}

File: MileageStats.Web/Scripts/mobile-debug/vehicle.js
/*  
Copyright Microsoft Corporation

Licensed under the Apache License, Version 2.0 (the "License"); you may not
use this file except in compliance with the License. You may obtain a copy of
the License at 

http://www.apache.org/licenses/LICENSE-2.0 

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED 
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE, 
MERCHANTABLITY OR NON-INFRINGEMENT. 

See the Apache 2 License for the specific language governing permissions and
limitations under the License. */

(function (mstats) {
	var ctr = function (require) {

	    var $ = require('$'),
	        window = require('window'),
            confirmPrompt = require('confirm'),
            formSubmitter = require('formSubmitter');

		var urlPattern = '/Vehicle/:vehicleId/Details';

		function postrender(model, el, context) {

			formSubmitter.attach(el, function (response) {
				window.location.hash = urlPattern.replace(':vehicleId', response.Model.VehicleId);
			});

			formSubmitter.attach(el, function (response) {
				window.location.hash = '/';
			}, '#DeleteVehicleForm');

			el.find('#DeleteVehicleButton').click(function (event) {
				var answer = confirmPrompt("Are you sure you want to delete this vehicle and all of it's related data?");
				if (!answer) {
					event.preventDefault();
				}
			});

			var $vehicleEditForm = el.find('#vehicleEditForm'),
                $yearSelect = el.find('#Year'),
                $makeSelect = el.find('#MakeName'),
                $modelSelect = el.find('#ModelName'),
                makesUrl = $vehicleEditForm.data('makes-url'),
                modelsUrl = $vehicleEditForm.data('models-url');

			$yearSelect.change(function () {
				$.post(makesUrl,
                { year: $yearSelect.val() },
                function (data) {
                	$makeSelect.children().not(':first').remove();
                	$modelSelect.children().not(':first').remove();
                	updateList(data, $makeSelect);
                });
			});

			$makeSelect.change(function () {
				$.post(modelsUrl,
                { year: $yearSelect.val(), make: $makeSelect.val() },
                function (data) {
                	$modelSelect.children().not(':first').remove();
                	updateList(data, $modelSelect);
                });
			});
		};

		function updateList(data, $selectList) {
			$.each(data,
              function (key, value) {
              	$selectList.append(
                    $('<option></option>')
                    .attr('value', value)
                    .text(value)
                    );
              });
		};

		return {
			postrender: postrender
		};
	};
	mstats.vehicleEdit = ctr;
	mstats.vehicleAdd = ctr;
} (this.mstats = this.mstats || {}));
